Implementing Internationalization with SvelteKit and Directus (#131)
directus-labs/guest-authoring: multilingual-news-app-using-sveltekit-with-directus/index.md

---
title: 'Build a Multilingual News app using Sveltekit with Directus i18n API'
description: 'This tutorial will guide you through building a multilingual news application using Sveltekit and Directus i18n API. Sveltekit is a modern frontend framework that enables fast and efficient development, while Directus i18n API provides a simple way to manage multilingual content. You will learn how to create a news app that supports multiple languages, fetch and display news articles, and manage translations using Directus i18n API. By the end of this tutorial, you will have a fully functional news app that can be easily extended and customized to meet your needs..'
author:
  name: 'Ekekenta Clinton'
  avatar_file_name: './ekekenta-clinton.png'
---

## Introduction
Svelte is a new JavaScript framework for building user interfaces for web applications. It offer a better way for handling UI updates when compared with React, Angular, or Vue.js. In this tutorial, you'll learn how to build a multilingual news application using Sveltekit with Directus i18n API.

> ## Before You Start
> You will need:
> - [Node.js v20.11.1](https://nodejs.org/) or later
> - A code editor on your computer.
> - Create and login to your [Directus Cloud account](https://directus.cloud/) and have install the Directus SDK `npm install @directus/sdk`.
> - Some knowledge of React and Svelte.
> 
> The code for this tutorial is available on my [GitHub repository](https://github.com/icode247/directus-i18n-app); feel free to clone it to follow along.

Reviewer: ```suggestion
## Before You Start

You will need:

- [Node.js v20.11.1](https://nodejs.org/) or later.
- A code editor on your computer.
- A Directus project - follow our [quickstart guide](https://docs.directus.io/getting-started/quickstart) if you don't already have one.
- Some knowledge of React and Svelte.

The code for this tutorial is available on my [GitHub repository](https://github.com/icode247/directus-i18n-app).
```

## The Concept of i18n

i18n is the process of creating applications that support multiple languages. The phrase i18n represents the first and the last letters of the word **internationalization**. The number 18 denotes the number of letters between the first and the last letter. The main goal of i18n is to conveniently create applications that support multiple languages or locations without having to redesign the application to add those languages.

## Understanding Directus i18n API

Directus is an open-source headless Content Management System that provides a robust API for managing content. It offers an i18n API that enables the internationalization and localization of content within the system.

The Directus i18n API allows you to manage translations for various content types, such as articles, products, or any other data stored in the Directus system.

> With the Directus i18n API, you can build multilingual applications or websites that serve localized content to users based on their preferred language or region. 

Reviewer: Remove this - I don't think it adds much that hasn't already been said.

> ## Setting up a new project.
> 
> Let's start by setting up a new Svelte project using [degit](https://github.com/Rich-Harris/degit), by running the command below:
> 
> ```
>  npx degit sveltejs/template directus-i18n-app
> ```
> 
> The above command will scalfold a new Svelte project with the required dependencies. 
> 
> Next, change directory into the project folder and install the dependencies.
> 
> ```
> cd directus-i18n-app && npm install
> ```
> 
> Then install the **svelte-routing** package as a dev dependency:
> ```
> npm i -D svelte-routing
> ```

Reviewer: ````suggestion
Let's start by setting up a new Svelte project using the [degit scaffolding tool](https://github.com/Rich-Harris/degit):

```
npx degit sveltejs/template directus-i18n-app
```

Install the required dependencies:

```
cd directus-i18n-app 
npm install
npm install @directus/sdk
npm install --save-dev svelte-routing
```

In the `src/libs` directory, create a `directus.js` file to create and export a Directus SDK instance:

```
import { createDirectus, rest } from '@directus/sdk';
const directus = createDirectus('<YOUR DIRECTUS URL>').with(rest());
export default directus;
```

Be sure to provide your real project URL.
````

This should be a continuation of 'Before You Start'. As you can see, we've brought all of the dependency installation into a single instruction to make this more concise. The SDK install has also been moved here as it's a related step.

I've also added the boring boilerplate of setting up the SDK to this section.

> ## Designing the Data Model in Directus
> In your Directus dashboard, navigate to Settings -> Data Model and create a new collections for `news` and`languages`. Use the schema below for the articles collection:
> 
> ```
> news
> - id (Primary Key Field, Type: Auto-incremented Integer )
> - author (Type: String, Type: Manually entered string )
> - cover (Type: Image)
> ```
> ![Creating Directus news collection](./creating-directus-news-collection.png)
> 
> 
> For the languages collection, use the schema below:
> 
> ```
> languages
> 
> - code (Primary Key Field, Type: Manually entered string )
> - name (Type: String, Inferface: Input)
> - direction (Type: String, Interface: Dropdown, Default Value: ltr)
> ```
> 
> The `diretion` field is to enable support for languages that read right to left.
> ![Creating Dirctus languages collection](./creating-directus-language-collection.png)
> 
> 
> To enable content translation in your **news** collection, update the collection to create a new field for `translations`. The translation field will create a one-to-many relationship with the `languages` collection. So select the `languages` collection as the **Language Collection Field**, `name` field for the **Language Indicator Field** and the `direction` field for the **Language Direction Field**.
> 
> 
> ![Creating Directus translations collection](./creating-news-translation-collection.png)
> 
> 
> Once you Save, a new collection named `news_translations` will be created for you. In the `news_translations` collection, we'll add the fields that needs to translated. Update the collection to add the schema below:
> 
> ```
> news_translations
> 
> - title (Type: String, Interface: Input )
> - slug (Type: String, Interface: Input)
> - body (Type: Text, Inferface: WYSIWYG)
> ```
> ![Adding new fields to Directus news collection](./news-translation.png)
> 
> Now, in the Language collection, create new entries for the language you'd want your contents to be translated to. You'll add entries for German, English (US) and French. 
> ![Creating new entries in the languages collections](./languages-entries.png)
> 
> Then create some entries in your News collection.
> 
> ![Creating new entries in the news collections](./news-translation-entries.png)
> 
> 
> Now you can write your contents in different translations (German, French and English).
> 
> Lastly allow the Public role to read the `news`, `languages` and `news_translations` collections in the Access Control settings.
> 
> 

Reviewer: ```suggestion
## Designing a Data Model

In the Directus Data Studio, navigate to Settings -> Data Model and create a new collection called `news`:

- `id` (Primary Key Field, Type: Auto-incremented Integer)
- `author` (Type: String, Interface: Input)
- `cover` (Type: Image)

Create a collection called `languages`:

- `code` (Primary Key Field, Type: Manually entered string )
- `name` (Type: String, Interface: Input)
- `direction` (Type: String, Interface: Dropdown, Options: `ltr` and `rtl`. Default Value: `ltr`)

The `direction` field enables support for languages that read right to left.

To enable content translation in your `news` collection, add a new One to Many `translations` field with the `languages` collection. Select `name` as the **Language Indicator Field** and `direction` as the **Language Direction Field**.

![Creating Directus translations collection](./creating-news-translation-collection.png)

Once you save, a new collection named `news_translations` will be created for you. In the `news_translations` collection, you will add the fields that need translations. 

Add the following fields to the `news_translations` collection:

- `title` (Type: String, Interface: Input)
- `slug` (Type: String, Interface: Input)
- `body` (Type: Text, Interface: WYSIWYG)

Add each language you want to support as items in the `languages` collection. 

![Creating new entries in the languages collections](./languages-entries.png)

The item page for the `news` collection now includes a translations interface.

![Creating new entries in the news collections](./news-translation-entries.png)

Allow the Public role to read the `news`, `languages` and `news_translations` collections in the Access Control settings to ensure the frontend can access these fields.
```

I've made some suggestions to make this much more succinct, but I want to point out that you're shifting from second person (you) to first person plural (our) pronouns. This is good to keep an eye on as you write more content - stay consistent. I've changed them to second person for you.

## Building the News App Frontend with SvelteKit
In your Svelte project, create a `directus.js` file to initialize Dirctus SDK in your project:

```
import { createDirectus, rest } from '@directus/sdk';

const directus = createDirectus('<YOUR DIRECTUS URL>').with(rest());

export default directus;

> ```

Reviewer: Remove
